File: dp-python/dp-algorithms/zero_one_knapsack.py
def zero_one_knapsack(value_array, weight_array, total_capacity, index):
    if total_capacity == 0 or index == 0:
        return 0

    if weight_array[index - 1] <= total_capacity:
        first_answer = value_array[index - 1] + zero_one_knapsack(value_array, weight_array,
                                                                  total_capacity - weight_array[index - 1], index - 1)
        second_answer = zero_one_knapsack(value_array, weight_array, total_capacity, index - 1)
        return max(first_answer, second_answer)
    else:
        return zero_one_knapsack(value_array, weight_array, total_capacity, index - 1)


def zero_one_knapsack_memoization(value_array, weight_array, total_capacity, index, dp):
    if total_capacity == 0 or index == 0:
        return 0

    if dp[index][total_capacity] != -1:
        return dp[index][total_capacity]

    if weight_array[index - 1] <= total_capacity:
        first_answer = value_array[index - 1] + zero_one_knapsack_memoization(value_array, weight_array,
                                                                              total_capacity - weight_array[index - 1],
                                                                              index - 1, dp)
        second_answer = zero_one_knapsack_memoization(value_array, weight_array, total_capacity, index - 1, dp)
        dp[index][total_capacity] = max(first_answer, second_answer)
    else:
        dp[index][total_capacity] = zero_one_knapsack_memoization(value_array, weight_array, total_capacity, index - 1,
                                                                  dp)

    return dp[index][total_capacity]


def zero_one_knapsack_tabulation(value_array, weight_array, total_capacity):
    dp_table = [[0 for _ in range(total_capacity + 1)] for _ in range(len(value_array) + 1)]

    for i in range(1, len(value_array) + 1):
        value_ = value_array[i - 1]
        weight_ = weight_array[i - 1]
        for j in range(1, total_capacity + 1):
            if weight_ <= j:
                include = value_ + dp_table[i - 1][j - weight_]
                exclude = dp_table[i - 1][j]
                dp_table[i][j] = max(include, exclude)
            else:
                dp_table[i][j] = dp_table[i - 1][j]

    return dp_table[len(value_array)][total_capacity]


def target_subset(arr, target_sum):
    dp_table = [[False for _ in range(target_sum + 1)] for _ in range(len(arr) + 1)]
    for i in range(len(arr) + 1):
        dp_table[i][0] = True

    for i in range(1, len(arr) + 1):
        for j in range(1, target_sum + 1):
            value_ = arr[i - 1]
            if value_ <= j and dp_table[i - 1][j - value_]:
                dp_table[i][j] = True
            elif dp_table[i - 1][j]:
                dp_table[i][j] = True
    return dp_table[len(arr)][target_sum]


if __name__ == '__main__':
    value, weight, capacity = [15, 14, 10, 45, 30], [2, 5, 1, 3, 4], 7
    print(zero_one_knapsack(value, weight, capacity, len(value)))
    two_dimension_array = [[-1 for i in range(capacity + 1)] for _ in range(len(value) + 1)]
    print(zero_one_knapsack_memoization(value, weight, capacity, len(value), two_dimension_array))
    print(zero_one_knapsack_tabulation(value, weight, capacity))
    print(target_subset([4, 2, 7, 1, 3], 10))  # O(n * sum)

File: dp-java/src/introductiondynamicprogramming/ZeroOneKnapsack.java
package introductiondynamicprogramming;

import java.util.Arrays;

public class ZeroOneKnapsack {
    static int knapsack(int[] value, int[] weight, int totalCapacity, int n){
        if (totalCapacity == 0 || n == 0){
            return 0;
        }

        if (weight[n-1] <= totalCapacity){
            int firstAnswer = value[n-1] + knapsack(value, weight, totalCapacity-weight[n-1], n-1); // Include
            int secondAnswer = knapsack(value, weight, totalCapacity, n-1); //Exclude
            return Math.max(firstAnswer, secondAnswer);
        }else { // Not valid
            return knapsack(value, weight, totalCapacity, n-1);
        }
    }

    static int knapsackMemoization(int[] value, int[] weight, int totalCapacity, int n, int[][] dp){
        if (totalCapacity == 0 || n == 0){
            return 0;
        }

        if(dp[n][totalCapacity] != -1){
            return dp[n][totalCapacity];
        }

        if (weight[n-1] <= totalCapacity){
            int firstAnswer = value[n-1] + knapsackMemoization(value, weight, totalCapacity-weight[n-1], n-1, dp); // Include
            int secondAnswer = knapsackMemoization(value, weight, totalCapacity, n-1, dp); //Exclude
            dp[n][totalCapacity] = Math.max(firstAnswer, secondAnswer);
        }else { // Not valid
            dp[n][totalCapacity] = knapsackMemoization(value, weight, totalCapacity, n-1, dp);
        }
        return dp[n][totalCapacity];
    }

    static int knapsackTabulation(int[] value, int[] weight, int capacity){
        int n = value.length;
        int[][] dp = new int[n+1][capacity+1];

        for (int i=1; i<n+1; i++){
            for(int j=1; j<capacity+1; j++){
                int v = value[i-1];
                int w = weight[i-1];
                if (w <= j){ // Valid
                    int includeProfit = v + dp[i-1][j-w];
                    int excludeProfit = dp[i-1][j];
                    dp[i][j] = Math.max(includeProfit, excludeProfit);
                } else {
                    int excludeProfit = dp[i-1][j];
                    dp[i][j] = excludeProfit;
                }
            }
        }
        return dp[n][capacity];

    }

    static boolean targetSubSubset(int[] arr, int sum){ // O(n * sum)
        int n = arr.length;
        boolean[][] dp = new boolean[n+1][sum+1];
        for(int i=0; i<n+1; i++){
            dp[i][0] = true;
        }

        for (int i=1; i<n+1; i++){
            for (int j=1; j<sum+1; j++){
                int v = arr[i-1];
                //Include
                if (v <= j && dp[i - 1][j - v]){
                    dp[i][j] = true;
                }

                //Exclude
                else if (dp[i-1][j]) {
                    dp[i][j] = true;
                }

            }
        }

        for (boolean[] i: dp){
            System.out.println(Arrays.toString(i));
        }

        return dp[n][sum];
    }

    public static void main(String[] args) {
        int[] value = {15, 14, 10, 45, 30};
        int[] weight = {2, 5, 1, 3, 4};
        int totalCapacity = 7;

        int[][] dp = new int[value.length+1][totalCapacity+1];
        for(int i=0; i<dp.length; i++){
            for (int j=0; j<dp[0].length; j++){
                dp[i][j] = -1;
            }
        }

        System.out.println(knapsack(value, weight, totalCapacity, value.length));
        System.out.println(knapsackMemoization(value, weight, totalCapacity, value.length, dp)); // O(n*totalCapacity)
        System.out.println(knapsackTabulation(value, weight, totalCapacity));

        int[] arr = {4, 2, 7, 1, 3};
        int sum = 10;

        System.out.println(targetSubSubset(arr, sum));


    }
}
